### PR TITLE
Add df.empty property

### PR DESCRIPTION
**Pull request by rockwellw**
- [ ] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by TomAugspurger** [comment]
We’ll need to be careful here. Getting the length of the DataFrame is expensive since it requires computation.

Also I suspect that checking df.empty is typically followed by a bool, so we maybe don’t want a delayed result here.

**Comment by TomAugspurger** [comment]
One more comment on the implementation, `DataFrame.empty` should indicate whether either dimension has length 0. This just checks whether the rows are length 0.

**Comment by rockwellw** [comment]
Agreed on both points @TomAugspurger .  From what I could tell, there doesn't seem to be any metadata that would allow a quick answer to this question.  Do you have any suggestions?
And if there's no quick way to calculate this at the moment, do you still think its worth adding the property so that the behavior more closely matches pandas?

**Comment by TomAugspurger** [comment]
I'm not sure. It's generally expected that accessing a property is inexpensive. Potentially triggering a read from disk to check that would be surprising.

This may be a case where it's appropriate to explicitly differ from pandas, either by not implementing DataFrame.empty, or by providing / documenting something like `df.map_blocks(lambda x: x.empty).compute().all()` as an alternative.

**Comment by mrocklin** [comment]
What about returning a delayed value?  Something like ...

```python
def empty(self):
    return self.size == 0
```

That way people could do something like `df.empty.compute()` to get the full value. 

However, if they include this within an if statement we would want to make sure that to err.  It looks like Dask dataframe scalars don't err out currently when we call `bool` on them.  We should probably change this?

**Comment by CMCDragonkai** [comment]
I have a situation where I want to lazily return something different if the dataframe is empty. But this condition must also be lazy.

Right now if I use `len(df.index) == 0` this performs a computation ahead of time.

What is the right way to get a "lazy" length check, and then a lazy condition.

It  seems like dask requires some conditional combinator that is lazy.

Maybe there is a way to combine length checks with `dask.delayed`?

**Comment by TomAugspurger** [comment]
You can use `.map_partitions(len)` and (lazily) sum those to get the length
of the dataframe.

**Comment by mrocklin** [comment]
Perhaps also the `size` property?

**Comment by TomAugspurger** [comment]
> What about returning a delayed value?

I'm a bit hesitant to do that. In my experience, `.empty` will almost always be used with a boolean check. With a Dask DataFrame, checking the sizes of the two dimensions have very different costs. I think we should encourage the user to think about this carefully.

So I think my preferred change would be 

1. Clear docs on Why dask.dataframe doesn't implement `.empty`
2. Possibly, we implement a `.empty` that explicitly raises a `NotImplementedError` with a clear message indicating what to do, depending on whether you're checking for no rows vs. no columns (or both).

**Comment by mrocklin** [comment]
Sounds good to me.  I prefer 2 over 1

**Comment by TomAugspurger** [comment]
@rockwellw does the earlier discussion make sense to you? Do you have time to implement a `NotImplementedError` on DataFrame.empty to raise with a nice error message explaining the situation?

**Comment by rockwellw** [comment]
@TomAugspurger yeah that makes sense.  Thanks for sorting that out!  I'll update the PR soon.

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -3292,6 +3292,14 @@ def __len__(self):
         else:
             return len(s)
 
+    @property
+    def empty(self):
+        raise NotImplementedError(
+            "Checking the dimensions of a Dask DataFrame can be expensive. "
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: The only thing I would clarify here is that only checking whether there are any rows is expensive. Checking the number of columns is fast.

**Comment by TomAugspurger** [comment]
Thanks @rockwellw!

**Pull request merged**